core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java

```diff
@@ -0,0 +1,253 @@
+            // only resets the size() to 0 which controls the index where subsequent calls to add()
+            // will store new values, but does not modify the positions array
+            int[] positions = positionsList.elements();
+            positionsList.clear();
+
+            if (bucketSize == page.getPositionCount()) {
+                // entire page will be sent to this writer.
+                return buffers.get(bucket);
+            }
+            Page pageSplit = page.copyPositions(positions, 0, bucketSize);
+            sendPageToPartition(buffers.get(bucket), pageSplit);
+        }
+
+        // No single writer receives the entire input page
+        return null;
+    }
+
+    private int getNextPartitionWriter(int partition, long[] partitionPhysicalWrittenBytes)
+    {
+        PartitionScalingInfo scalingInfo = partitionScalingInfos[partition];
+        int maxWriterCount = scalingInfo.getMaxWriterCount();
+
+        // Scale up writers for a partition when current buffer memory utilization is more than 50% of the
+        // maximum and physical written bytes is greater than writerMinSize.
+        // This also mean that we won't scale local writers if the writing speed can cope up
+        // with incoming data. In another word, buffer utilization is below 50%.
+        if (maxWriterCount <= (buffers.size() - initialWriterCount) && memoryManager.getBufferedBytes() >= maxBufferedBytes / 2) {
+            long physicalWrittenBytes = partition < partitionPhysicalWrittenBytes.length ? partitionPhysicalWrittenBytes[partition] : 0;
+            if (physicalWrittenBytes >= writerMinSize * maxWriterCount) {
+                scalingInfo.incrementWriterCount();
+                log.debug("Increased task writer count to " + (maxWriterCount + 1) + " for partition " + partition);
+            }
+        }
+
+        // Find the next writer in a round-robin fashion.
+        int nextWriterCount = scalingInfo.getNextWriterCount();
+
+        // If the next writer is equal to zero, then we will choose between initial writers using a simple mod
+        // between the partition id and initialWriterCount.
+        // If the next writer is above zero which means scaling has happened, then we will choose among the unused writers.
+        return nextWriterCount == 0 ? partition & initialWriterMask : initialWriterCount + nextWriterCount - 1;
+    }
+
+    private void sendPageToPartition(Consumer<PageReference> buffer, Page pageSplit)
+    {
+        PageReference pageReference = new PageReference(pageSplit, 1, onPageReleased);
+        memoryManager.updateMemoryUsage(pageReference.getRetainedSizeInBytes());
+        buffer.accept(pageReference);
+    }
+
+    @Override
+    public ListenableFuture<Void> waitForWriting()
+    {
+        return memoryManager.getNotFullFuture();
+    }
+
+    private static class PartitionScalingInfo
+    {
+        private int maxWriterCount = 1;
```

Comment:
nit: Maybe different variable name as it this points to how many writers where created and not a max we could have and both could mix together.